File: quickstart_guides/arrays_strings/python/length_of_longest_substring.py
"""
Title: Longest substring without repeating characters.

Problem:
    Given a string, find the length of the longest substring without repeating
    characters.

Execution: python length_of_longest_substring.py
"""
import unittest


def length_of_longest_substring(s: str) -> int:
    """Given a string, determine the length of the longest substring."""
    last, res, st = dict(), 0, 0
    for i, v in enumerate(s):
        if v not in last or last[v] < st:
            res = max(res, i - st + 1)
        else:
            st = last[v] + 1
        last[v] = i
    return res


class TestLengthOfLongestSubstring(unittest.TestCase):
    """Unit tests for length_of_longest_substring."""

    def test_1(self):
        """Test for string: abcabcbb."""
        self.assertEqual(length_of_longest_substring("abcabcbb"), 3)
        print("Explanation: The answer is 'abc', with the length of 3.")

    def test_2(self):
        """Test for string: bbbbb."""
        self.assertEqual(length_of_longest_substring("bbbbb"), 1)
        print("Explanation: The answer is 'b', with the length of 1.")

    def test_3(self):
        """Test for string: pwwkew."""
        self.assertEqual(length_of_longest_substring("pwwkew"), 3)
        print("Explanation: The answer is 'wke', with the length 3.")
        print("Note that the answer must be a substring, 'pwke' ")
        print("is a subsequence and not a substring.")


if __name__ == "__main__":
    unittest.main()

File: quickstart_guides/arrays_strings/java/LengthOfLongestSubstring.java
/*
 *   Title: Longest substring without repeating characters.
 *
 *   Problem
 *      Given a string, find the length of the longest substring without
 *      repeating characters.
 *
 *   Execution: javac LengthOfLongestSubstring.java && java LengthOfLongestSubstring
 */

import java.util.HashMap;

public class LengthOfLongestSubstring {
    public static long lengthOfLongestSubstring(String s) {
        int n = s.length(), ans = 0;

        // Current index of character.
        int[] index = new int[128];

        // Attempt to extend the range.
        int i = 0;
        for (int j = 0; j < n; j++) {
            i = Math.max(index[s.charAt(j)], i);
            ans = Math.max(ans, j - 1 + 1);
            index[s.charAt(j)] = j + 1;
        }
        return ans;
    }

    public static void main(String[] args) {
        assert lengthOfLongestSubstring("abcabcbb") == 3;
        System.out.println("Explanation: The answer is 'abc', with the length of 3.");

        assert lengthOfLongestSubstring("bbbbb") == 1;
        System.out.println("Explanation: The answer is 'b', with the length of 1.");

        assert lengthOfLongestSubstring("pwwkew") == 3;
        System.out.println("Explanation: The answer is 'wke', with the length 3. Note that the answer must be a substring, 'pwke' is a subsequence and not a substring.");

        System.out.println("Passed all test cases");
    }
    
}
